src/Stratis.Bitcoin/AddressIndexing/AddressIndexer.cs

```diff
@@ -0,0 +1,284 @@
+                    this.logger.LogDebug("Tip was not found, initializing with genesis.");
+
+                    this.data = new AddressIndexerData()
+                    {
+                        TipHash = this.network.GenesisHash.ToString(),
+                        AddressIndexDatas = new List<AddressIndexData>()
+                    };
+                    this.dataStore.Insert(this.data);
+                }
+
+                this.blockReceivedQueue = new AsyncQueue<KeyValuePair<bool, ChainedHeaderBlock>>(this.OnEnqueueAsync);
+
+                // Subscribe to events.
+                this.blockConnectedSubscription = this.signals.Subscribe<BlockConnected>(blockConnectedData =>
+                {
+                    while (this.blockReceivedQueue.Count > 100)
+                    {
+                        this.logger.LogWarning("Address indexing is slowing down the consensus.");
+                        Thread.Sleep(5000);
+                    }
+
+                    this.blockReceivedQueue.Enqueue(new KeyValuePair<bool, ChainedHeaderBlock>(true, blockConnectedData.ConnectedBlock));
+                });
+
+                this.blockDisconnectedSubscription = this.signals.Subscribe<BlockDisconnected>(blockDisconnectedData =>
+                {
+                    this.blockReceivedQueue.Enqueue(new KeyValuePair<bool, ChainedHeaderBlock>(false, blockDisconnectedData.DisconnectedBlock));
+                });
+
+                this.nodeStats.RegisterStats(this.AddComponentStats, StatsType.Component, 400);
+            }
+        }
+
+        private Task OnEnqueueAsync(KeyValuePair<bool, ChainedHeaderBlock> item, CancellationToken cancellationtoken)
+        {
+            try
+            {
+                this.ProcessBlockAddedOrRemoved(item);
+            }
+            catch (Exception ex)
+            {
+                this.logger.LogError(ex.ToString());
+                throw;
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private void AddComponentStats(StringBuilder benchLog)
+        {
+            benchLog.AppendLine();
+            benchLog.AppendLine("======AddressIndexer======");
+
+            benchLog.AppendLine($"Unprocessed blocks: {this.blockReceivedQueue.Count}");
+        }
+
+        /// <summary>Processes block that was added or removed from consensus chain.</summary>
+        private void ProcessBlockAddedOrRemoved(KeyValuePair<bool, ChainedHeaderBlock> item)
+        {
+            // Make sure it's on top of the tip.
+            bool blockAdded = item.Key;
+
+            if ((blockAdded && item.Value.ChainedHeader.Header.HashPrevBlock.ToString() != this.data.TipHash) ||
+                (!blockAdded && item.Value.ChainedHeader.HashBlock.ToString() != this.data.TipHash))
+            {
+                const string message = "TransactionIndexer is in inconsistent state. This can happen if you've enabled txindex on an already synced or partially synced node. " +
```

Comment:
So `txindex` also enables `addressindex`? is this correct? this is even more of a reason why this should be its own feature or part of store.

Reply:
Correct

Reply:
I think there should be a separate flag for addressindex otherwise just by enabling trxindex you take so much more space.
Do we have stats of te size of hte db on STRAT  mainchain?

Reply:
>Do we have stats of te size of hte db on STRAT mainchain?

No. Just indexing the txes takes a while so I never synced the full chain. 
I will though.

Reply:
>I think there should be a separate flag for addressindex otherwise just by enabling trxindex you take so much more space.

Agreed. I'll add in the next PR